sdb/commands/stacks.py

```diff
@@ -0,0 +1,255 @@
+        return (-1, 0)
+
+    def validate_args(self, prog: drgn.Program,
+                      args: argparse.Namespace) -> None:
+        if args.function:
+            try:
+                func = prog[args.function]
+            except KeyError:
+                raise sdb.CommandError(
+                    self.name,
+                    "symbol '{:s}' does not exist".format(args.function))
+            if func.type_.kind != drgn.TypeKind.FUNCTION:
+                raise sdb.CommandError(
+                    self.name, "'{:s}' is not a function".format(args.function))
+
+        task_states = Stacks.TASK_STATES.values()
+        if args.tstate and not args.tstate in task_states:
+            raise sdb.CommandError(
+                self.name,
+                "'{:s}' is not a valid task state (acceptable states: {:s})".
+                format(args.tstate, ", ".join(task_states)))
+
+        if args.module and Stacks.find_module_memory_segment(
+                self.prog, args.module)[0] == -1:
+            raise sdb.CommandError(
+                self.name,
+                "module '{:s}' doesn't exist or isn't currently loaded".format(
+                    args.module))
+
+    def call(self, objs: Iterable[drgn.Object]) -> Iterable[drgn.Object]:
+        # pylint: disable=too-many-locals
+        # pylint: disable=too-many-branches
+
+        #
+        # As the exception explains the code that follows this statement only
+        # works for linux kernel targets (crash dumps or live systems). When
+        # support for userland is added we can factor the kernel code that
+        # follows into its own function and switch to the correct codepath
+        # depending on the target.
+        #
+        if not self.prog.flags & drgn.ProgramFlags.IS_LINUX_KERNEL:
+            raise sdb.CommandError(self.name,
+                                   "userland targets are not supported yet")
+        self.validate_args(self.prog, self.args)
+
+        mod_start, mod_end = -1, -1
+        if self.args.module:
+            mod_start, mod_size = Stacks.find_module_memory_segment(
+                self.prog, self.args.module)
+            assert mod_start != -1
+            mod_end = mod_start + mod_size
+
+        print("{:<18s} {:<16s} {:6s}".format("TASK_STRUCT", "STATE", "COUNT"))
+        print("=" * 42)
+
+        stack_map = {}
+        stack_counter = Counter()
+        task_aggr = defaultdict(list)
+
+        #
+        # For now this iterates over the initial namespace only. The drgn
+        # API does provide support for iterating over a specific namespace
+        # so functiality can be added where the `stacks` command is passed
+        # namespace `drgn.Object`s through the pipeline or a namespace ID
+        # through `argparse` and iterates over the tasks of those. This
+        # functionality has been left as future work.
+        #
+        for task in for_each_task(self.prog):
```

Comment:
I'd prefer skipping this. I think we should first get a better understanding of how namespaces and tasks are bundled together before explicitly writing a comment about wanting to print X by default.

Reply:
I do think that this is the behavior we will want eventually. I think this command is most useful for getting an overall picture of the activity on a system from the perspective of the kernel, across all threads. If someone runs `stacks -m zfs` for instance, it would be confusing to leave out some of the threads that are in ZFS code doing writes because they are in other containers. You could end up in situations where, for example, you can see a bunch of threads waiting on a lock, but not see anyone holding the lock.

That said, if you'd rather leave this out until we investigate more thoroughly, I think that's reasonable.

Reply:
That's fair! I'll incorporate it

Reply:
I looked at the kernel code and asked around in the LKDC slack to verify. The way that pid namespaces are structured is hierarchical with the `init_pid_ns` at level 0, and any `pid_namespace`s after that in subsequent levels (max is 32) in a tree-like structure of parent-children relationships.

So using `for_each_task()` above passing `drgn.Progam()` which according to the `drgn` API passes the initial namespace, we get the behavior that we want (it picks up every `task_struct`, each of which may have a pid in other namespaces)